packages/esm-patient-chart-app/src/visit-header/workspaces.ts

```diff
@@ -0,0 +1,203 @@
+
+import { getExtensionRegistration } from '@openmrs/esm-framework/src/internal';
+import _i18n from 'i18next';
+
+type WorkspaceWindowState = 'minimized' | 'maximized' | 'hidden' | 'reopened' | 'normal';
+
+export interface Prompt {
+  title: string;
+  body: string;
+  confirmText?: string;
+  onConfirm(): void;
+  cancelText?: string;
+}
+
+export interface WorkspaceStoreState {
+  patientUuid: string | null;
+  openWorkspaces: Array<OpenWorkspace>;
+  prompt: Prompt | null;
+}
+
+export interface OpenWorkspace extends WorkspaceRegistration {
+  additionalProps: object;
+  closeWorkspace(promptBeforeClosing?: boolean): void;
+  closeWorkspace(): void;
+  promptBeforeClosing(testFcn: () => boolean): void;
+}
+
+export interface WorkspaceRegistration {
+  name: string;
+  title: string;
+  load(): Promise<any>;
+  type?: string;
+  preferredWindowSize?: WorkspaceWindowState;
+}
+
+let registeredWorkspaces: Record<string, WorkspaceRegistration> = {};
+
+/**
+ * This exists for compatibility with the old way of registering
+ * workspaces (as extensions).
+ *
+ * @param name of the workspace
+ */
+function getWorkspaceRegistration(name: string): WorkspaceRegistration {
+  if (registeredWorkspaces[name]) {
+    return registeredWorkspaces[name];
+  } else {
+    const workspaceExtension = getExtensionRegistration(name);
+    if (workspaceExtension) {
+      return {
+        name: workspaceExtension.name,
+        title: getTitleFromExtension(workspaceExtension),
+        preferredWindowSize: workspaceExtension.meta?.screenSize ?? 'normal',
+        load: workspaceExtension.load,
+        type: workspaceExtension.meta?.type ?? 'form',
+      };
+    } else {
+      throw new Error(`No workspace named '${name}' has been registered.`);
+    }
+  }
+}
+
+function getTitleFromExtension(ext: ExtensionRegistration) {
+  const title = ext?.meta?.title;
+  if (typeof title === 'string') {
+    return title;
+  } else if (title && typeof title === 'object') {
+    return translateFrom(ext.moduleName, title.key, title.default);
+  }
+  return ext.name;
+}
+
+/**
+ * Given a workspace specified by its name:
+ *
+ * - If no workspaces are open, or if no other workspaces with the same type are open,
+ *   it will be opened and focused.
+ * - If a workspace with the same name is already open, it will be displayed/focused,
+ *   if it was not already.
+ * - If another workspace with the same type is open, a confirmation modal will pop up a
+ *   warning about closing the currently open workspace.
+ *
+ * Note that this function just manipulates the workspace store. The UI logic based on
+ * the workspace store is contained in esm-patient-chart-app.
+ *
+ * @param name The name of the workspace to launch
+ * @param additionalProps Props to pass to the workspace component being launched
+ */
+export function launchPatientWorkspace(name: string, additionalProps?: object) {
```

Comment:
I think I'm missing why this file is necessary.. Don't we already have `launchPatientWorkspaces` from [esm-patient-common-lib](https://github.com/openmrs/openmrs-esm-patient-chart/blob/main/packages/esm-patient-common-lib/src/workspaces/workspaces.ts) ? If this file is redundant let's remove it

Reply:
Yeah I have realized that, I will delete the workspace.ts and use the workspace from the common lib. Thanks for the catch.